modules/thermalsensor_generic.c

```diff
@@ -0,0 +1,1451 @@
+        *list = g_slist_prepend(*list, object);
+    }
+
+    return object;
+}
+
+/** Locate/add sensor object in list of sensors
+ *
+ * @param list  pointer to the head of linked list
+ * @param name  name of the sensor to find
+ *
+ * @return sensor object
+ */
+static thermal_sensor_generic_t *
+tsg_objects_add_sensor(GSList **list, const char *name)
+{
+    thermal_object_t *object = tsg_objects_add_object(list, name);
+    return thermal_sensor_generic_from_object(object);
+}
+
+/** Convert thermal limit name thermal status enum value
+ *
+ * @param key key string
+ *
+ * @return thermal status, or -1 on failure
+ */
+static THERMAL_STATUS
+tsg_objects_parse_level(const char *key)
+{
+    static const struct
+    {
+        const char *key;
+        int         val;
+    } lut[] = {
+        { "Low",     THERMAL_STATUS_LOW     },
+        { "Normal",  THERMAL_STATUS_NORMAL  },
+        { "Warning", THERMAL_STATUS_WARNING },
+        { "Alert",   THERMAL_STATUS_ALERT   },
+        { "Fatal",   THERMAL_STATUS_FATAL   },
+        { "Invalid", THERMAL_STATUS_INVALID },
+        { 0,         -1 }
+    };
+
+    THERMAL_STATUS status = -1;
+
+    for( int i = 0; ; ++i ) {
+        if( lut[i].key && strcmp(lut[i].key, key) )
+            continue;
+
+        status = lut[i].val;
+        break;
+    }
+
+    return status;
+}
+
+/** Validate and register a list of thermal objects
+ *
+ * @param list  pointer to the head of a linked list
+ */
+static void
+tsg_objects_register_all(GSList **list)
+{
+    for( GSList *item = *list; item; item = item->next ) {
+        thermal_object_t *object = item->data;
+
+        if( !object )
+            continue;
+
+        thermal_sensor_generic_t *sensor =
+            thermal_sensor_generic_from_object(object);
+
+        if( !sensor )
+            continue;
+
+        if( !thermal_sensor_generic_is_valid(sensor) ) {
+            dsme_log(LOG_ERR, PFIX"%s: %s",
+                     thermal_sensor_generic_get_name(sensor),
+                     "sensor config is not valid");
+        }
+        else if( !thermal_sensor_generic_enable_sensor(sensor, true) ) {
+            dsme_log(LOG_ERR, PFIX"%s: %s",
+                     thermal_sensor_generic_get_name(sensor),
+                     "sensor could not be enabled");
+        }
+        else if( !thermal_sensor_generic_read_sensor(sensor) ) {
+            dsme_log(LOG_ERR, PFIX"%s: %s",
+                     thermal_sensor_generic_get_name(sensor),
+                     "sensor could not be read");
+        }
+        else {
+            thermal_manager_register_object(object);
+            continue;
+        }
+
+        item->data = 0;
+        thermal_object_delete(object);
+    }
+}
+
+/** Read configuration file and generate thermal objects
+ *
+ * @param list    pointer to the head of a linked list
+ * @param config  path to configuration file
+ */
+static void
+tsg_objects_read_config(GSList **list, const char *config)
```

Comment:
Procedural logic with no asynchronous events. To me it is the simplest part of the new code.

While having unit tests does not hurt, new config files are needed roughly once / new device port. The whole functionality needs to be manually verified anyway. I nowadays get allergic reactions if "simple" is mentioned too close to "unit test" -> I'll create separate task for this (we need protection against breaking already existing files).